vllm/worker/tt_worker.py

```diff
@@ -9,16 +9,164 @@
+        kv_cache_shape = (num_blocks, self.num_kv_heads, self.block_size, self.head_size)
+        kv_cache: List[torch.Tensor] = []
+        num_layers = self.num_attention_layers 
+        # num_layers = 1 # TODO: Debugging for 1 layer 
+        if device == "cpu":
+            for _ in range(num_layers):
+                # null block in CpuGpuBlockAllocator requires at least that
+                # block to be zeroed-out.
+                # Zero-initialize CPU cache
+                cache_k = torch.zeros(kv_cache_shape,
+                                      dtype=self.dtype,
+                                      device=device)
+                cache_v = torch.zeros(kv_cache_shape,
+                                      dtype=self.dtype,
+                                      device=device)
+                kv_cache.append([cache_k, cache_v])
+        else:
+            for _ in range(num_layers):
+                cache_k = torch.zeros(kv_cache_shape, dtype=self.dtype)
+                cache_v = torch.zeros(kv_cache_shape, dtype=self.dtype)
+                
+                kv_tt = [ttnn.as_tensor(
+                    lp,
+                    device=self.device_config.device,
+                    # TODO: this could be ShardTensorToMesh, removing need for init to know about TP=8. Could affect other calculations which use self.num_kv_heads, though.
+                    mesh_mapper=ReplicateTensorToMesh(self.device_config.device),
+                    layout=ttnn.TILE_LAYOUT,
+                    memory_config=ttnn.DRAM_MEMORY_CONFIG,
+                    dtype=ttnn.bfloat8_b
+                ) for lp in (cache_k, cache_v)]
+                
+                kv_cache.append(kv_tt)
+        return kv_cache
+
+    def swap_in(self, src_to_dst: torch.Tensor) -> None:
+        raise NotImplementedError
+        # for i in range(self.num_attention_layers):
+        #     self.attn_backend.swap_blocks(self.cpu_cache[i], self.gpu_cache[i],
+        #                                   src_to_dst)
+
+    def swap_out(self, src_to_dst: torch.Tensor) -> None:
+        raise NotImplementedError
+        # for i in range(self.num_attention_layers):
+        #     self.attn_backend.swap_blocks(self.gpu_cache[i], self.cpu_cache[i],
+        #                                   src_to_dst)
+
+    def copy(self, src_to_dsts: torch.Tensor) -> None:
+        raise NotImplementedError
+        # self.attn_backend.copy_blocks(self.gpu_cache, src_to_dsts)
+
+    @staticmethod
+    def get_cache_block_size(
+        cache_config: CacheConfig,
+        model_config: ModelConfig,
+        parallel_config: ParallelConfig,
+    ) -> int:
+        head_size = model_config.get_head_size()
+        num_heads = model_config.get_num_kv_heads(parallel_config)
+        num_heads //= 8 # TODO: Make general without using TP=8?
+        num_attention_layers = model_config.get_num_attention_layers(
+            parallel_config)
```

Comment:
can use self.head_size, self.num_kv_heads, self.num_attention_layers here

Reply:
Unfortunately I can't because this is a static method

Reply:
Yeah sorry I meant to say as well we should remove the @staticmethod decorator

Reply:
When we allow vLLM to determine how many blocks to allocate based on available memory, I think this has to be a static function because it is called before the class is instantiated into an object

Reply:
Okay I see, in that case could we use a separate function to get head_size, num_kv_heads, num_attention_layers, dtype (or maybe just only num_kv_heads since `num_heads //= 8` is what we will likely change later)? Would just like to de-duplicate the code in __init__ and the static function so that we avoid the two from diverging in the future.

Reply:
Good point, I put this calculation into a static function. Now the suspicious calculation is only in one place